### ISSUE TITLE
Interesting errors

### ISSUE THREAD

**Issue by lulco** [efabrica-team/phpstan-latte#193]
- [x] Invalid identifier name "" (fixed in https://github.com/efabrica-team/phpstan-latte/pull/197)
- [x] Expression of type Expr_Variable cannot be evaluated (fixed in https://github.com/efabrica-team/phpstan-latte/pull/218)
- [x] Method Nette\Forms\Controls\Checkbox::getControlPart() invoked with 1 parameter, 0 required. (ignored, see https://github.com/efabrica-team/phpstan-latte/pull/226)
- [x] Method Nette\Forms\Controls\Checkbox::getLabelPart() invoked with 1 parameter, 0 required. (ignored, see https://github.com/efabrica-team/phpstan-latte/pull/226)
- [x] Variable $item might not be defined. - if some variable is created e.g. in foreach probably it is passed to included template - see https://github.com/spaze/michalspacek.cz/pull/63#issuecomment-1374050290 (considered valid)
- [x] Cannot access offset 'videoEmbed' on Nette\Forms\Container|false. (fixed https://github.com/efabrica-team/phpstan-latte/pull/249)

Collected on https://github.com/spaze/michalspacek.cz repository

```
 ------ ---------------------------------------------------------------------------------------------------------------- 
  Line   app/Www/Presenters/templates/Talks/talk.latte rendered from MichalSpacekCz\Www\Presenters\TalksPresenter::talk  
 ------ ---------------------------------------------------------------------------------------------------------------- 
  -1     Invalid identifier name ""                                                                                      
 ------ ---------------------------------------------------------------------------------------------------------------- 

 ------ ------------------------------------------------------------------------------------------------------------------------------ 
  Line   app/Www/Presenters/templates/Trainings/success.latte rendered from MichalSpacekCz\Www\Presenters\TrainingsPresenter::success  
 ------ ------------------------------------------------------------------------------------------------------------------------------ 
  -1     Expression of type Expr_Variable cannot be evaluated                                                                          
 ------ ------------------------------------------------------------------------------------------------------------------------------

 ------ ------------------------------------------------------------------------------------------------------------------------ 
  Line   app/Admin/Presenters/templates/Talks/slides.latte rendered from MichalSpacekCz\Admin\Presenters\TalksPresenter::slides  
 ------ ------------------------------------------------------------------------------------------------------------------------ 
  20     Form field with name "number" probably does not exist.                                                                  
  63     Form field with name "number" probably does not exist.                                                                  
  105    If condition is always true.                                                                                            
  105    Method Nette\Forms\Controls\Checkbox::getControlPart() invoked with 1 parameter, 0 required.                            
  105    Method Nette\Forms\Controls\Checkbox::getLabelPart() invoked with 1 parameter, 0 required.                              
 ------ ------------------------------------------------------------------------------------------------------------------------ 

 ------ ------------------------------------------------------------------------------------------------------------------------------------ 
  Line   app/Admin/Presenters/templates/Interviews/default.latte rendered from MichalSpacekCz\Admin\Presenters\InterviewsPresenter::default  
 ------ ------------------------------------------------------------------------------------------------------------------------------------ 
  1      Variable $item might not be defined.                                                                                                
  2      Variable $item might not be defined.                                                                                                
  3      Variable $item might not be defined.                                                                                                
  4      Variable $item might not be defined.                                                                                                
  5      Variable $item might not be defined.                                                                                                
  11     Variable $form might not be defined.                                                                                                
  37     Variable $interview on left side of ?? always exists and is not nullable.                                                           
  39     Cannot access offset 'videoEmbed' on Nette\Forms\Container|false.                                                                   
  42     Cannot access offset 'sourceName' on Nette\Forms\Container|false.                                                                   
  45     Cannot access offset 'sourceHref' on Nette\Forms\Container|false.                                                                   
  48     Cannot access offset 'submit' on Nette\Forms\Container|false.                                                                       
 ------ ------------------------------------------------------------------------------------------------------------------------------------ 
```

**Comment by lulco** [comment]
@spaze @MartinMystikJonas Any idea where the error `Invalid identifier name ""` is from? I did't find it in nette, latte, phpstan-latte and also in phpstan…

**Comment by lulco** [comment]
"Expression of type Expr_Variable cannot be evaluated" is probably caused by some wrong node replacement. Need to debug it.

**Comment by MartinMystikJonas** [comment]
> @spaze @MartinMystikJonas Any idea where the error `Invalid identifier name ""` is from? I did't find it in nette, latte, phpstan-latte and also in phpstan…

IMHO it is from creation of Identifier node in nikic/php-parser - it is propagated exception message

**Comment by MartinMystikJonas** [comment]
> xpression of type Expr_Variable cannot be evaluated

This is thrown by PhpParser\ConstExprEvaluator when no fallback callback is set. But no idea where ConstExprEvaluator  is used that way because we use it only in ValueResolver with fallback.

**Comment by lulco** [comment]
> > @spaze @MartinMystikJonas Any idea where the error `Invalid identifier name ""` is from? I did't find it in nette, latte, phpstan-latte and also in phpstan…
> 
> IMHO it is from creation of Identifier node in nikic/php-parser - it is propagated exception message

Caused by BetterReflection when no mapping was set. Fixed in https://github.com/efabrica-team/phpstan-latte/pull/197

**Comment by MartinMystikJonas** [comment]
About "Variable $item might not be defined" I think this is valid error. Varaible is missinng from block definition. And that block could be used for example in child template where it would cause error.

**Comment by lulco** [comment]
> About "Variable $item might not be defined" I think this is valid error. Varaible is missinng from block definition. And that block could be used for example in child template where it would cause error.

Well, it works in application says @spaze. Maybe because there is something like get_declared_variables() and they are sent to the included template.

**Comment by MartinMystikJonas** [comment]
> > About "Variable $item might not be defined" I think this is valid error. Varaible is missinng from block definition. And that block could be used for example in child template where it would cause error.
> 
> Well, it works in application says @spaze. Maybe because there is something like get_declared_variables() and they are sent to the included template.

It is not template include but block include. Blocks inherits all variables from call context. But we can never know all places from where this block is called. Define of block should specify parameters.

**Comment by spaze** [comment]
> About "Variable $item might not be defined" I think this is valid error. Varaible is missinng from block definition. And that block could be used for example in child template where it would cause error.

Agree, it is a valid error and I'm planning to resolve those reports. I think it can, and should be fixed by something _explicit_ like

```
{include #block, item => $item}
```

But I haven't tested it like that, yet.

**Comment by MartinMystikJonas** [comment]
You need to define parameter in `{define blockName Type $item}`. Also note that in Latte all defined block parameters are always nullable. So you also need to add check for null.

**Comment by spaze** [comment]
Ah, thanks, that's a good info!

**Comment by lulco** [comment]
@spaze can you please paste compiled code for this template Admin/Presenters/templates/Talks/default.latte  ? 

Error should be different and transformed to more readable version.

```
Line   Admin/Presenters/templates/Talks/default.latte rendered from           
[53](https://github.com/spaze/michalspacek.cz/actions/runs/3897801662/jobs/6655848292#step:6:54)
         MichalSpacekCz\Admin\Presenters\InterviewsPresenter::interview         
[54](https://github.com/spaze/michalspacek.cz/actions/runs/3897801662/jobs/6655848292#step:6:55)
         included in Admin/Presenters/templates/Interviews/default.latte        
[55](https://github.com/spaze/michalspacek.cz/actions/runs/3897801662/jobs/6655848292#step:6:56)
         included in Admin/Presenters/templates/Interviews/interview.latte      
[56](https://github.com/spaze/michalspacek.cz/actions/runs/3897801662/jobs/6655848292#step:6:57)
 ------ ----------------------------------------------------------------------- 
[57](https://github.com/spaze/michalspacek.cz/actions/runs/3897801662/jobs/6655848292#step:6:58)
  -1     Method                                                                 
[58](https://github.com/spaze/michalspacek.cz/actions/runs/3897801662/jobs/6655848292#step:6:59)
         MichalSpacekCz\Admin\Presenters\InterviewsPresenter::slidesWithParame  
[59](https://github.com/spaze/michalspacek.cz/actions/runs/3897801662/jobs/6655848292#step:6:60)
         tersMissingCorrespondingMethod not found
```

**Comment by spaze** [comment]
> @spaze can you please paste compiled code for this template Admin/Presenters/templates/Talks/default.latte ?
> 
> Error should be different and transformed to more readable version.

Yeah, that's indeed an interesting error :-) Moved to #233 and linked the source there.

**Comment by spaze** [comment]
> About "Variable $item might not be defined" I think this is valid error.

Btw, these errors have disappeared now without me doing anything (I'm now running 0.0.3, was 0.0.2 when this issue was created)

**Comment by lulco** [comment]
> > About "Variable $item might not be defined" I think this is valid error.
> 
> Btw, these errors have disappeared now without me doing anything (I'm now running 0.0.3, was 0.0.2 when this issue was created)

Did you rewrite the code? Or maybe it is not analyzed now for some reason  :)

**Comment by spaze** [comment]
> Did you rewrite the code?

No, I haven't ¯\\_(ツ)_/¯

**Comment by lulco** [comment]
Can you add some debug into it? For example `{php \PHPStan\dumpType($item)}`

**Comment by MartinMystikJonas** [comment]
> > About "Variable $item might not be defined" I think this is valid error.
> 
> Btw, these errors have disappeared now without me doing anything (I'm now running 0.0.3, was 0.0.2 when this issue was created)

That is weird. Can you share compiled template from /tmp/phpstan-latte?

**Comment by spaze** [comment]
> > > About "Variable $item might not be defined" I think this is valid error.
> > 
> > 
> > Btw, these errors have disappeared now without me doing anything (I'm now running 0.0.3, was 0.0.2 when this issue was created)
> 
> That is weird. Can you share compiled template from /tmp/phpstan-latte?

Here they are https://gist.github.com/spaze/ae214b8c1f1e5ba5816889d09e09cf80 (originally for #233)

**Comment by MartinMystikJonas** [comment]
> > > > About "Variable $item might not be defined" I think this is valid error.
> > > 
> > > 
> > > Btw, these errors have disappeared now without me doing anything (I'm now running 0.0.3, was 0.0.2 when this issue was created)
> > 
> > 
> > That is weird. Can you share compiled template from /tmp/phpstan-latte?
> 
> Here they are https://gist.github.com/spaze/ae214b8c1f1e5ba5816889d09e09cf80 (originally for #233)

That looks corrent and PHPStan should report missing variable because it is not defined there. Can you add `{php \PHPStan\dumpType($item)}` at the start of block itemInterview? And try what it will dump into PHPStan errors?

I wonder what PHPStan thinks about that variable. Because it is not defined anywhere.

Did you maybe changed your PHPStan config somehow? Or could it be change in new PHPStan version?

**Comment by spaze** [comment]
> I wonder what PHPStan thinks about that variable. Because it is not defined anywhere.

Oh! 😅 Thanks for this particular hint. Of course I'm stupid: I'm running phpstan-latte with a different separate PHPStan config so I can test templates in parallel with running "regular" PHPStan (which runs much faster). And the config file has `customRulesetUsed: true` and no level specified which essentially means PHPStan is using level 0 - which means that possibly undefined variables are not reported. _Of course!_ 😅

Compare these two, same file, different levels:
1. https://phpstan.org/r/ab942671-2e2a-4db7-beb0-3f9a67e3109b level 0 (_what undefined variable?_)
2. https://phpstan.org/r/287aad3c-89f6-49af-bb1b-532e158b99b4 level 1, contains "Variable $item might not be defined."

I think it would be nice to add a line or two to docs saying that maybe you want to run PHPStan with rule level at least 1 :-) I'll now increase the level and see what I need to do, I can then create the PR (will file an issue for now).

**Comment by spaze** [comment]
Just to answer this (although see above, found the issue why "Variable $item might not be defined" suddenly stopped being reported)

> Can you add some debug into it? For example `{php \PHPStan\dumpType($item)}`

This is the output:
```
------ ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
  Line   Admin/Presenters/templates/Interviews/default.latte rendered from MichalSpacekCz\Admin\Presenters\InterviewsPresenter::interview included in Admin/Presenters/templates/Interviews/interview.latte
 ------ ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
  6      Dumped type: mixed
 ------ ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

 ------ --------------------------------------------------------------------------------------------------------------------------------
  Line   Admin/Presenters/templates/Interviews/default.latte rendered from MichalSpacekCz\Admin\Presenters\InterviewsPresenter::default
 ------ --------------------------------------------------------------------------------------------------------------------------------
  6      Dumped type: mixed
 ------ --------------------------------------------------------------------------------------------------------------------------------


 [ERROR] Found 2 errors

```

**Comment by lulco** [comment]
Last issue, undefined variable is considered valid, so I'm closing this issue. If you’re find anything else, please create new issue.
